app/code/community/Bolt/Boltpay/Block/Checkout/Boltpay.php

```diff
@@ -511,6 +511,39 @@ public function getCartURL()
         return $this->boltHelper()->getMagentoUrl('checkout/cart');
     }
 
+    /**
+     * Returns the One Page or Multi-Page checkout Publishable key depending on the requested
+     * page and which key has been set in the admin.  Cart and Product pages prioritize the multi-step key
+     * and falls back to the payment only key.  All other pages prioritize the payment only key and falls back
+     * to the multi-step key.
+     *
+     * @return string   The publishable key to be used on this page
+     * @throws Bolt_Boltpay_BoltException when neither a multi-step nor a payment-only publishable key is configured
+     */
+    public function getPublishableKeyForThisPage() {
+        $routeName = $this->getRequest()->getRouteName();
+        $controllerName = $this->getRequest()->getControllerName();
+
+        $multiStepPublishableKey = $this->getPublishableKey('multi-page');
+        $paymentOnlyPublishableKey = $this->getPublishableKey('one-page');
+
+        $thisPagesPublishableKey =
+            $controllerName === 'cart'
+            || ($routeName === 'catalog' && $controllerName === 'product')
+                ? $multiStepPublishableKey
+                : $paymentOnlyPublishableKey
+        ;
+
+        if (!$multiStepPublishableKey && !$paymentOnlyPublishableKey) {
+            $noPublishableKeyException = new Bolt_Boltpay_BoltException("No publishable key has been configured.");
+            $this->boltHelper()->logException($noPublishableKeyException);
+            $this->boltHelper()->notifyException($noPublishableKeyException, [], 'error');
+            throw $noPublishableKeyException;
+        }
+
+        return $thisPagesPublishableKey ?: max($multiStepPublishableKey, $paymentOnlyPublishableKey);
```

Comment:
using max function has no sense. This is like random choice.

Reply:
I will humor these comments as it may prove valuable for you and for others of our team who may wish to speed learn PHP and our M1 plugin.

First, @oldDriver, it is not that the code does not make sense, it is that you do not understand it.  In that situation, there is no problem asking for explanation and please do, as it is preferred.  

The code does concisely what it is suppose to do. Here we can only have three cases:

1.) Both publishable keys are configured and the preferred one was set to `$thisPagesPublishableKey`;
2.) Only the preferred publishable key was configured and is set to `$thisPagesPublishableKey`.
3.) Only the non-preferred publishable key was configured and `$thisPagesPublishableKey` is an empty string,`''`, or `null`

If both keys were either null or empty, the  `$noPublishableKeyException` would have been thrown.

In case #1 and #2, `$thisPagesPublishableKey` is returned with the preferred key.

In case #3, we know that the non-preferred key is a string and what we now want to return and we also know that the preferred string set to `$thisPagesPublishableKey` is an empty string or `null`.  The `max` function will now compare the keys.  Because PHP type juggles, it will now convert `null` to and empty string to compare the same types.

https://www.php.net/manual/en/language.types.string.php#language.types.string.casting

The `max` documentation is very clear about its usage in string comparisons. 

![image](https://user-images.githubusercontent.com/973856/70265066-ee14d800-1799-11ea-80f8-d2df11e25ad1.png)

https://www.php.net/manual/en/function.max.php

So, the end result will the non-empty string will be returned, which we know to be the non-preferred key, which is concisely what we are after.

Reply:
@mccottry just one dummy question: Could merchant start with Multistep key and later add OnePage step?
What will be in this case?
Great thanks for the "php.net" link! never see before

Reply:
@oldDriver, @matt-thomason, @ryan-timothy-albert 

> @mccottry just one dummy question: Could merchant start with Multistep key and later add OnePage step?
> What will be in this case?

In this case, on the shopping cart page, and all product pages, the multi-step key is presented.  For all other cases, it will be assumed that it is a one-page checkout and so that key will be presented.  Also note, that in current use, this only is used for track.js and not connect.js.